docs/NetEbpfExtMultiAttach.md

```diff
@@ -0,0 +1,126 @@
+required data structures changes are described below.
+
+Even though multiple programs will attach with same "attach params", there will be a single instance of the
+corresponding WFP filters added. This means there will be one-to-many mapping from filter context to client context.
+The hook provider will now contain a list of filter contexts, which in turn will have list (or an array) of the clients
+corresponding to those attach params.
+
+Whenever a new eBPF program is attached, following will be the new flow:
+1. Check the list of current filter contexts in hook provider, to see if a filter context with same attach params
+already exists. If an existing filter context is found, the new client is added to the list of client contexts in the
+existing filter context.
+2. If an existing filter context does not exist, create a new filter context and a new client context.
+3. Configure corresponding WFP filters.
+4. Add the new client context to the list of client contexts in the filter context.
+5. Add the filter context in the list of filter contexts in hook provider context.
+
+**Multiple program invocation flow**
+
+Whenever WFP callout is invoked  based on a matching filter, the callout driver gets pointer to filter context. From
+the filter context, the callout gets the list of all the attached clients. The callout loops over all the clients and
+invokes them. The *out* context of one program invocation becomes the *in* context of the next program. In the chain of
+invocation, if any program **drops** the packet, the chain breaks there, and the callout returns the verdict to WFP.
+
+
+## Synchronization
+In the new design, access to the list of clients in the filter context and the list of filter contexts in the provider
+context needs to be synchronized. The 2 flows that need synchronization specifically are:
+1. Program invocation
+2. Client attach / detach callback.
+
+To provide synchronization, the hook provider will maintain a *dispatch* level RW lock to synchronize access to these
+lists. (A dispatch level lock is chosen here as the WFP callouts can be invoked at either PASSIVE_LEVEL or DISPATCH_LEVEL).
+1. Attach and detach callbacks flow will acquire this lock in exclusive mode.
+2. Program invocation flow acquires this lock in shared mode.
+
+This approach allows multiple program invocations at the same time, as well as protecting the list when a client is
+being attached / detached. Client / attach operations are expected to be much less frequent than program invocations.
+This approach eliminates any need for rundown reference for the client context. Since the list of client contexts for a
+filter context is synchronized via a lock, it is always guaranteed that in the client detach flow, if the dispatch level
+lock is acquired, there is no program invocation in progress for that program. Also, once the client context is removed
```

Comment:
"if the dispatch level *write lock* is acquired" ....

Reply:
Lines 89-90 and 114 don't use the term "read" or "write" lock, they use "exclusive" or "shared".